pkg/kubelet/cadvisor/cadvisor_linux.go

```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
 	return cadvisorClient, nil
 }
 
+// New takes a memory storage and returns a new manager.
+func newCadvisor(memoryCache *memory.InMemoryCache, sysfs sysfs.SysFs, maxHousekeepingInterval time.Duration, allowDynamicHousekeeping bool, includedMetricsSet container.MetricSet, collectorHttpClient *http.Client, rawContainerCgroupPathPrefixWhiteList []string) (manager.Manager, error) {
+	if memoryCache == nil {
+		return nil, fmt.Errorf("manager requires memory storage")
+	}
+
+	// Detect the container we are running on.
+	selfContainer, err := cgroups.GetOwnCgroupPath("cpu")
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("cAdvisor running in container: %q", selfContainer)
+
+	var (
+		dockerStatus info.DockerStatus
+		rktPath      string
+	)
+	docker.SetTimeout(dockerClientTimeout)
+	// Try to connect to docker indefinitely on startup.
+	dockerStatus = retryDockerStatus()
+
+	crioClient, err := crio.Client()
+	if err != nil {
+		return nil, err
+	}
+	crioInfo, err := crioClient.Info()
+	if err != nil {
+		klog.V(5).Infof("CRI-O not connected: %v", err)
+	}
```

Comment:
@dashpole @yujuhong @derekwaynecarr @mrunalp 

Per [this comment](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cadvisor/util.go#L66-L72):
```
// CRI integrations should get container metrics via CRI. Docker
// uses the built-in cadvisor to gather such metrics on Linux for
// historical reasons.
// cri-o relies on cadvisor as a temporary workaround. The code should
// be removed. Related issue:
// https://github.com/kubernetes/kubernetes/issues/51798
// UsingLegacyCadvisorStats returns true if container stats are provided by cadvisor instead of through the CRI
```

Gathering crio stats from cadvisor is a temporary workaround. The referenced issue was auto-closed. What are the conditions for removing the cri-o dependency here?

@dims - If we can't remove it now, can you file a new issue with the criteria identified, and add a comment here to remove this once that issue is resolved.

Reply:
@mrunalp can correct me, but I don't think CRI-O implements the stats interface yet.

Reply:
From the code it looks like support is there: https://github.com/cri-o/cri-o/blob/master/server/container_stats_list.go

Reply:
So i have reopened https://github.com/kubernetes/kubernetes/issues/51798

and will add a comment above `crioInfo, err := crioClient.Info()` similar to what's above `UsingLegacyCadvisorStats` method